streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,644 @@
+
+            for (final TopicPartition partition : polledRecords.partitions()) {
+                final ChangelogMetadata changelogMetadata = changelogs.get(partition);
+                if (changelogMetadata == null) {
+                    throw new IllegalStateException("The corresponding changelog restorer for " + partition +
+                        " does not exist, this should not happen.");
+                }
+                if (changelogMetadata.changelogState != ChangelogState.RESTORING) {
+                    throw new IllegalStateException("The corresponding changelog restorer for " + partition +
+                        " has already transited to completed state, this should not happen.");
+                }
+
+                final List<ConsumerRecord<byte[], byte[]>> records = polledRecords.records(partition);
+                final Iterator<ConsumerRecord<byte[], byte[]>> iterator = records.iterator();
+                while (iterator.hasNext()) {
+                    final ConsumerRecord<byte[], byte[]> record = iterator.next();
+                    if (record.key() == null) {
+                        log.warn("Read changelog record with null key from changelog {} at offset {}, " +
+                            "skipping it for restoration", changelogMetadata.changelogPartition, record.offset());
+                        iterator.remove();
+                    }
+                }
+
+                // TODO: we always try to restore as a batch when some records are accumulated, which may result in
+                //       small batches; this can be optimized in the future, e.g. wait longer for larger batches.
+                if (!records.isEmpty()) {
+                    if (changelogMetadata.bufferedRecords != null && !changelogMetadata.bufferedRecords.isEmpty()) {
+                        changelogMetadata.bufferedRecords.addAll(records);
+                    } else {
+                        changelogMetadata.bufferedRecords = records;
+                    }
+
+                    restoreChangelog(changelogMetadata);
+                }
             }
-        });
-
-        final Iterator<TopicPartition> iter = initializable.iterator();
-        while (iter.hasNext()) {
-            final TopicPartition topicPartition = iter.next();
-            final Long restoreOffset = restoreToOffsets.get(topicPartition);
-            final StateRestorer restorer = stateRestorers.get(topicPartition);
-
-            if (restorer.checkpoint() >= restoreOffset) {
-                restorer.setRestoredOffset(restorer.checkpoint());
-                iter.remove();
-                completedRestorers.add(topicPartition);
-            } else if (restoreOffset == 0) {
-                restorer.setRestoredOffset(0);
-                iter.remove();
-                completedRestorers.add(topicPartition);
+        }
+    }
+
+    /**
+     * returns a flag whether offset limit caused not all records restored
```

Comment:
This method doesn't return anything, need to update docs

Reply:
Ack, good catch.